### PR TITLE
credits page

### LINKED ISSUE #22: type error on macros
when i remove code of conduct i get error on macros.py

```python
src/iqdb_tagger/main.pymkdocs build -s✗ Building documentation (1)
  > mkdocs build -s
  INFO     -  [macros] - Macros arguments: {'module_name': 'docs/macros',
              'modules': [], 'include_dir': '', 'include_yaml': [],
              'j2_block_start_string': '', 'j2_block_end_string': '',
              'j2_variable_start_string': '', 'j2_variable_end_string': '',
              'verbose': False}
  INFO     -  [macros] - Found local Python module 'docs/macros' in:
              /mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger
  INFO     -  [macros] - Found external Python module 'docs/macros' in:
              /mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger
  INFO     -  [macros] - Extra filters (module): ['pretty']
  INFO     -  Cleaning site directory
  INFO     -  Building documentation to directory:
              /mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger/site
  WARNING  -  A relative path to 'code_of_conduct.md' is included in the 'nav'
              configuration, which is not found in the documentation files
  Package(s) not found: astunparse, atomicwrites, cached-property, pywin32-ctypes
  INFO     -  [macros] - ERROR # _Macro Rendering Error_

              **TypeError**: tuple indices must be integers or slices, not str


              ```
              Traceback (most recent call last):
                File
              "/home/r3r/.cache/pypoetry/virtualenvs/iqdb-tagger-C28AsU-2-py3.9/lib/python3.9/site-packages/mkdocs_macros/plugin.py",
              line 434, in render
                  return md_template.render(**page_variables)
                File
              "/home/r3r/.cache/pypoetry/virtualenvs/iqdb-tagger-C28AsU-2-py3.9/lib/python3.9/site-packages/jinja2/environment.py",
              line 1090, in render
                  self.environment.handle_exception()
                File
              "/home/r3r/.cache/pypoetry/virtualenvs/iqdb-tagger-C28AsU-2-py3.9/lib/python3.9/site-packages/jinja2/environment.py",
              line 832, in handle_exception
                  reraise(*rewrite_traceback_stack(source=source))
                File
              "/home/r3r/.cache/pypoetry/virtualenvs/iqdb-tagger-C28AsU-2-py3.9/lib/python3.9/site-packages/jinja2/_compat.py",
              line 28, in reraise
                  raise value.with_traceback(tb)
                File "<template>", line 1, in top-level template code
                File
              "/mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger/docs/macros.py",
              line 81, in credits
                  return get_credits()
                File
              "/mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger/docs/macros.py",
              line 63, in get_credits
                  template_data = get_credits_data()
                File
              "/mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger/docs/macros.py",
              line 35, in get_credits_data
                  pkg = {_: pkg[_] for _ in ("name", "home-page")}  # type: ignore
                File
              "/mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger/docs/macros.py",
              line 35, in <dictcomp>
                  pkg = {_: pkg[_] for _ in ("name", "home-page")}  # type: ignore
              TypeError: tuple indices must be integers or slices, not str

              ```

  Aborted with 1 warnings in strict mode!
WPS332 Found walrus operator> poetry run mkdocs build -s
INFO     -  [macros] - Macros arguments: {'module_name': 'docs/macros', 'modules': [], 'include_dir': '', 'include_yaml': [], 'j2_block_start_string': '', 'j2_block_end_string': '',
            'j2_variable_start_string': '', 'j2_variable_end_string': '', 'verbose': False}
INFO     -  [macros] - Found local Python module 'docs/macros' in: /mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger
INFO     -  [macros] - Found external Python module 'docs/macros' in: /mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger
INFO     -  [macros] - Extra filters (module): ['pretty']
INFO     -  Cleaning site directory
INFO     -  Building documentation to directory: /mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger/site
WARNING  -  A relative path to 'code_of_conduct.md' is included in the 'nav' configuration, which is not found in the documentation files
Package(s) not found: astunparse, atomicwrites, cached-property, pywin32-ctypes
WARNING:root:tuple indices must be integers or slices, not str
Traceback (most recent call last):
  File "/mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger/docs/macros.py", line 37, in get_credits_data
    pkg = {_: pkg[_] for _ in ("name", "home-page")}  # type: ignore
  File "/mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger/docs/macros.py", line 37, in <dictcomp>
    pkg = {_: pkg[_] for _ in ("name", "home-page")}  # type: ignore
TypeError: tuple indices must be integers or slices, not str
INFO     -  [macros] - ERROR # _Macro Rendering Error_

            **KeyError**: 'name'


            ```
            Traceback (most recent call last):
              File "/home/r3r/.cache/pypoetry/virtualenvs/iqdb-tagger-C28AsU-2-py3.9/lib/python3.9/site-packages/mkdocs_macros/plugin.py", line 434, in render
                return md_template.render(**page_variables)
              File "/home/r3r/.cache/pypoetry/virtualenvs/iqdb-tagger-C28AsU-2-py3.9/lib/python3.9/site-packages/jinja2/environment.py", line 1090, in render
                self.environment.handle_exception()
              File "/home/r3r/.cache/pypoetry/virtualenvs/iqdb-tagger-C28AsU-2-py3.9/lib/python3.9/site-packages/jinja2/environment.py", line 832, in handle_exception
                reraise(*rewrite_traceback_stack(source=source))
              File "/home/r3r/.cache/pypoetry/virtualenvs/iqdb-tagger-C28AsU-2-py3.9/lib/python3.9/site-packages/jinja2/_compat.py", line 28, in reraise
                raise value.with_traceback(tb)
              File "<template>", line 1, in top-level template code
              File "/mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger/docs/macros.py", line 86, in credits
                return get_credits()
              File "/mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger/docs/macros.py", line 68, in get_credits
                template_data = get_credits_data()
              File "/mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger/docs/macros.py", line 41, in get_credits_data
                packages[pkg["name"].lower()] = pkg  # type: ignore
            KeyError: 'name'

            ```

Aborted with 1 warnings in strict mode!
  INFO     -  [macros] - Macros arguments: {'module_name': 'docs/macros',
              'modules': [], 'include_dir': '', 'include_yaml': [],
              'j2_block_start_string': '', 'j2_block_end_string': '',
              'j2_variable_start_string': '', 'j2_variable_end_string': '',
              'verbose': False}
  INFO     -  [macros] - Found local Python module 'docs/macros' in:
              /mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger
  INFO     -  [macros] - Found external Python module 'docs/macros' in:
              /mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger
  INFO     -  [macros] - Extra filters (module): ['pretty']
  INFO     -  Cleaning site directory
  INFO     -  Building documentation to directory:
              /mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger/site
  WARNING  -  A relative path to 'code_of_conduct.md' is included in the 'nav'
              configuration, which is not found in the documentation files
  Package(s) not found: astunparse, atomicwrites, cached-property, pywin32-ctypes

  Aborted with 1 warnings in strict mode!

make: *** [Makefile:45: check] Error 1
```
i tried to by pass that but got another error
```python
src/iqdb_tagger/main.pymkdocs build -s✗ Building documentation (1)
  > mkdocs build -s
  INFO     -  [macros] - Macros arguments: {'module_name': 'docs/macros',
              'modules': [], 'include_dir': '', 'include_yaml': [],
              'j2_block_start_string': '', 'j2_block_end_string': '',
              'j2_variable_start_string': '', 'j2_variable_end_string': '',
              'verbose': False}
  INFO     -  [macros] - Found local Python module 'docs/macros' in:
              /mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger
  INFO     -  [macros] - Found external Python module 'docs/macros' in:
              /mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger
  INFO     -  [macros] - Extra filters (module): ['pretty']
  INFO     -  Cleaning site directory
  INFO     -  Building documentation to directory:
              /mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger/site
  WARNING  -  A relative path to 'code_of_conduct.md' is included in the 'nav'
              configuration, which is not found in the documentation files
  Package(s) not found: astunparse, atomicwrites, cached-property, pywin32-ctypes
  INFO     -  [macros] - ERROR # _Macro Rendering Error_

              **TypeError**: tuple indices must be integers or slices, not str


              ```
              Traceback (most recent call last):
                File
              "/home/r3r/.cache/pypoetry/virtualenvs/iqdb-tagger-C28AsU-2-py3.9/lib/python3.9/site-packages/mkdocs_macros/plugin.py",
              line 434, in render
                  return md_template.render(**page_variables)
                File
              "/home/r3r/.cache/pypoetry/virtualenvs/iqdb-tagger-C28AsU-2-py3.9/lib/python3.9/site-packages/jinja2/environment.py",
              line 1090, in render
                  self.environment.handle_exception()
                File
              "/home/r3r/.cache/pypoetry/virtualenvs/iqdb-tagger-C28AsU-2-py3.9/lib/python3.9/site-packages/jinja2/environment.py",
              line 832, in handle_exception
                  reraise(*rewrite_traceback_stack(source=source))
                File
              "/home/r3r/.cache/pypoetry/virtualenvs/iqdb-tagger-C28AsU-2-py3.9/lib/python3.9/site-packages/jinja2/_compat.py",
              line 28, in reraise
                  raise value.with_traceback(tb)
                File "<template>", line 1, in top-level template code
                File
              "/mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger/docs/macros.py",
              line 81, in credits
                  return get_credits()
                File
              "/mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger/docs/macros.py",
              line 63, in get_credits
                  template_data = get_credits_data()
                File
              "/mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger/docs/macros.py",
              line 35, in get_credits_data
                  pkg = {_: pkg[_] for _ in ("name", "home-page")}  # type: ignore
                File
              "/mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger/docs/macros.py",
              line 35, in <dictcomp>
                  pkg = {_: pkg[_] for _ in ("name", "home-page")}  # type: ignore
              TypeError: tuple indices must be integers or slices, not str

              ```

  Aborted with 1 warnings in strict mode!
WPS332 Found walrus operator> poetry run mkdocs build -s
INFO     -  [macros] - Macros arguments: {'module_name': 'docs/macros', 'modules': [], 'include_dir': '', 'include_yaml': [], 'j2_block_start_string': '', 'j2_block_end_string': '',
            'j2_variable_start_string': '', 'j2_variable_end_string': '', 'verbose': False}
INFO     -  [macros] - Found local Python module 'docs/macros' in: /mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger
INFO     -  [macros] - Found external Python module 'docs/macros' in: /mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger
INFO     -  [macros] - Extra filters (module): ['pretty']
INFO     -  Cleaning site directory
INFO     -  Building documentation to directory: /mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger/site
WARNING  -  A relative path to 'code_of_conduct.md' is included in the 'nav' configuration, which is not found in the documentation files
Package(s) not found: astunparse, atomicwrites, cached-property, pywin32-ctypes
WARNING:root:tuple indices must be integers or slices, not str
Traceback (most recent call last):
  File "/mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger/docs/macros.py", line 37, in get_credits_data
    pkg = {_: pkg[_] for _ in ("name", "home-page")}  # type: ignore
  File "/mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger/docs/macros.py", line 37, in <dictcomp>
    pkg = {_: pkg[_] for _ in ("name", "home-page")}  # type: ignore
TypeError: tuple indices must be integers or slices, not str
INFO     -  [macros] - ERROR # _Macro Rendering Error_

            **KeyError**: 'name'


            ```
            Traceback (most recent call last):
              File "/home/r3r/.cache/pypoetry/virtualenvs/iqdb-tagger-C28AsU-2-py3.9/lib/python3.9/site-packages/mkdocs_macros/plugin.py", line 434, in render
                return md_template.render(**page_variables)
              File "/home/r3r/.cache/pypoetry/virtualenvs/iqdb-tagger-C28AsU-2-py3.9/lib/python3.9/site-packages/jinja2/environment.py", line 1090, in render
                self.environment.handle_exception()
              File "/home/r3r/.cache/pypoetry/virtualenvs/iqdb-tagger-C28AsU-2-py3.9/lib/python3.9/site-packages/jinja2/environment.py", line 832, in handle_exception
                reraise(*rewrite_traceback_stack(source=source))
              File "/home/r3r/.cache/pypoetry/virtualenvs/iqdb-tagger-C28AsU-2-py3.9/lib/python3.9/site-packages/jinja2/_compat.py", line 28, in reraise
                raise value.with_traceback(tb)
              File "<template>", line 1, in top-level template code
              File "/mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger/docs/macros.py", line 86, in credits
                return get_credits()
              File "/mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger/docs/macros.py", line 68, in get_credits
                template_data = get_credits_data()
              File "/mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger/docs/macros.py", line 41, in get_credits_data
                packages[pkg["name"].lower()] = pkg  # type: ignore
            KeyError: 'name'

            ```

Aborted with 1 warnings in strict mode!
  INFO     -  [macros] - Macros arguments: {'module_name': 'docs/macros',
              'modules': [], 'include_dir': '', 'include_yaml': [],
              'j2_block_start_string': '', 'j2_block_end_string': '',
              'j2_variable_start_string': '', 'j2_variable_end_string': '',
              'verbose': False}
  INFO     -  [macros] - Found local Python module 'docs/macros' in:
              /mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger
  INFO     -  [macros] - Found external Python module 'docs/macros' in:
              /mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger
  INFO     -  [macros] - Extra filters (module): ['pretty']
  INFO     -  Cleaning site directory
  INFO     -  Building documentation to directory:
              /mnt/ac54dceb-73a5-4f94-b52c-cb7a426c0f29/Documents/iqdb_tagger/site
  WARNING  -  A relative path to 'code_of_conduct.md' is included in the 'nav'
              configuration, which is not found in the documentation files
  Package(s) not found: astunparse, atomicwrites, cached-property, pywin32-ctypes

  Aborted with 1 warnings in strict mode!

make: *** [Makefile:45: check] Error 1
```
this is fix for python3.8
```diff
Author: Rachmadani Haryono <foreturiga@gmail.com>
Date:   Tue Dec 14 07:45:16 2021 +0800

    fix(macros): type error

diff --git a/project/docs/macros.py b/project/docs/macros.py
index 72262ad..2ccf826 100644
--- a/project/docs/macros.py
+++ b/project/docs/macros.py
@@ -4,6 +4,7 @@ import functools
 from itertools import chain
 from pathlib import Path

+import logging
 import httpx
 import toml
 from jinja2 import StrictUndefined
@@ -32,8 +33,13 @@ def get_credits_data() -> dict:

     packages = {}
     for pkg in search_packages_info(list(dependencies)):
-        pkg = {_: pkg[_] for _ in ("name", "home-page")}  # type: ignore
-        packages[pkg["name"].lower()] = pkg  # type: ignore
+        try:
+            pkg_data = {_: pkg[_] for _ in ("name", "home-page")}  # type: ignore
+        except TypeError as err:
+            logging.debug(err, exc_info=True)
+            pkg_data = {}
+        if pkg_name := pkg_data.get("name"):
+            packages[pkg_name.lower()] = pkg_data  # type: ignore

     # all packages might not be credited,
     # like the ones that are now part of the standard library
```

### PR DESCRIPTION
**Pull request by rachmadaniHaryono**
resolves #22

- template is based on pawamoy jinja template with table format from https://axgkl.github.io/docutools/about/credits/
- this may require minimum pip version to fix the linked issue

### CODE REVIEW

**Review by pawamoy — CHANGES_REQUESTED** [review]
Nice, thanks! Just a few quality things to fix (see CI)

File: project/docs/credits.md [hunk 1]
```diff
@@ -1 +0,0 @@
-{{ credits() }}
```
Comment (pawamoy) [inline project/docs/credits.md:1]: Hmm, how do you inject the credits page without this?

Comment (rachmadaniHaryono) [reply project/docs/credits.md:1]: https://github.com/pawamoy/copier-poetry/pull/24/commits/97cda4b8df951e45efca3ec100e21abd9c00ad95

**Comment by rachmadaniHaryono** [comment]
i think i misunderstand jinja and copier.
i thought file with jinja suffix on docs folder will generate new file when run `make docs` or `make docs-serve`

**Comment by pawamoy** [comment]
So, credits must be generated each time we build the docs (because dependencies can change regularly).
Without specific plugins, Markdown pages are not rendered through Jinja by MkDocs. That's what the macros plugin does.
But in our case, the credits template needs specific context, that MkDocs itself (or even macros) is not aware of.
That is why we used macros to "inject" the `get_credits` function into the global context, function that returns the contents of credits.

You tried to use a shortcut: that's a nice idea. But in your changes macros is injecting `get_credits_data`, while the credits template is trying to access the *return value* of `get_credits_data`. Maybe you could add a `{% set data = get_credits_data() %}` at the top of the template, and then prefix each Jinja variable with `data.`.

Or you could refactor to use the same logic as in the copier-pdm template, i.e. stop using macros and replace it with mkdocs-gen-files.

**Comment by rachmadaniHaryono** [comment]
> Maybe you could add a {% set data = get_credits_data() %} at the top of the template, and then prefix each Jinja variable with data..

 i think i will choose this one

File: project/docs/credits.md [hunk 2]
```diff
@@ -1 +1,25 @@
...
+<!-- Template repository: https://github.com/pawamoy/jinja-templates
+     Template path: credits.md
+-->
+
+# Credits
+{% set data = get_credits_data() %}
+These projects were used to build `{{ data['project_name'] }}`. **Thank you!**
```
Comment (pawamoy) [inline project/docs/credits.md]: Jinja allows you to access dictionary items as attributes:
```suggestion
These projects were used to build `{{ data.project_name }}`. **Thank you!**
```

Comment (rachmadaniHaryono) [reply project/docs/credits.md]: i will apply that to all data items

**Pull request merged**